nitro-protocol/contracts/VirtualPaymentApp.sol

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.7.6;
+pragma experimental ABIEncoderV2;
+
+import './interfaces/IForceMoveApp.sol';
+import './libraries/NitroUtils.sol';
+import './interfaces/INitroTypes.sol';
+
+/**
+ * @dev The VirtualPaymentApp contract complies with the ForceMoveApp interface and allows payments to be made virtually from Alice to Bob (participants[0] to participants[1]).
+ */
+contract VirtualPaymentApp is IForceMoveApp {
+    struct SignedVoucher {
+        bytes32 channelId;
+        uint256 amount;
+        INitroTypes.Signature signature; // signature on abi.encode(channelId,amount)
+    }
+
+    /**
+     * @notice Encodes application-specific rules for a particular ForceMove-compliant state channel.
+     * @dev Encodes application-specific rules for a particular ForceMove-compliant state channel.
+     * @param fixedPart Fixed part of the state channel.
+     * @param proof Array of recovered variable parts which constitutes a support proof for the candidate.
+     * @param candidate Recovered variable part the proof was supplied for.
+     */
+    function requireStateSupported(
+        FixedPart calldata fixedPart,
+        RecoveredVariablePart[] calldata proof,
+        RecoveredVariablePart calldata candidate
+    ) external pure override {
+        // This channel has only 4 states which can be supported:
+        // 0 prefund
+        // 1 postfund
+        // 2 redemption
+        // 3 final
+
+        // states 0,1,3 can be supported via unanimous consensus:
+
+        if (proof.length == 0) {
+            require(
+                NitroUtils.getClaimedSignersNum(candidate.signedBy) ==
+                    fixedPart.participants.length,
+                '!unanimous; |proof|=0'
+            );
+            if (candidate.variablePart.turnNum == 0) return; // prefund
+            if (candidate.variablePart.turnNum == 1) return; // postfund
+            if (candidate.variablePart.turnNum == 3) {
+                // final (note: there is a core protocol escape hatch for this, too, so it could be removed)
+                require(candidate.variablePart.isFinal, '!final; turnNum=3 && |proof|=0');
```

Comment:
If the `redemption` state was also a `final` state would that allow Bob to unilaterally clear the challenge? It would be nice if Bob submitting the voucher just closed the channel, since there's nothing Alice or Irene can do after that to change the outcome (without Bob's support).

Or does the `ForceMove` framework require everyone to sign the concluded state? [Comments](https://github.com/statechannels/go-nitro/blob/6678d15d9ce091b415883bb13e266f5b91379cce/nitro-protocol/contracts/ForceMove.sol#L118) seem to indicate that conclude requires final states but I don't actually see where that's enforced. 🤔

Reply:
It looks like this is already being discussed in https://github.com/statechannels/go-nitro/issues/813

Reply:
This is a good data point to feed into that discussion 👍